[connector/routing] Use inferred contexts (#44762)
open-telemetry/opentelemetry-collector-contrib: connector/routingconnector/README.md

Reviewer: ```suggestion
    - condition: resource.attributes["env"] == "prod"
      pipelines: [ logs/prod ]
    - condition: resource.attributes["env"] == "dev"
      pipelines: [ logs/dev ]
    - condition: resource.attributes["region"] == "east"
      pipelines: [ logs/east ]
    - condition: resource.attributes["region"] == "west"
      pipelines: [ logs/west ]
```

Adding `resource.` prefix.

The following settings are available:

- `table (required)`: the routing table for this connector.
- `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Use context-qualified paths (e.g., `resource.attributes["key"]`, `span.attributes["key"]`) to automatically infer the context (see [Context Inference](#context-inference)).

> - `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. Generally `condition` is preferred since it is more terse.

Reviewer: ```suggestion
- `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. Generally `condition` is preferred since it is more terse. May not be used with the deprecated `request` context.
```

It's still valid.

> - `table.context (optional)`: the [OTTL Context] in which the condition/statement will be evaluated. Supported values: `resource`, `span`, `metric`, `datapoint`, `log`. **Deprecated:** `request` — use `otelcol.client.metadata` or `otelcol.grpc.metadata` paths instead (see [Limitations](#limitations)). In most cases this field should be omitted; the context is inferred automatically from context-qualified paths. If specified, it takes precedence over inference.

Reviewer: ```suggestion
- `table.context (optional)`: the [OTTL Context](#supported-contexts) in which the condition/statement will be evaluated. **Deprecated:** `request` — use `otelcol.client.metadata` or `otelcol.grpc.metadata` paths instead (see [Limitations](#limitations)). In most cases this field should be omitted; the context is inferred automatically from context-qualified paths. If specified, it takes precedence over inference.
```

- `default_pipelines (optional)`: contains the list of pipelines to use when a record does not meet any of specified conditions.
- `error_mode (optional)`: determines how errors returned from OTTL statements are handled. Valid values are `propagate`, `ignore` and `silent`. If `ignore` or `silent` is used and a statement's condition has an error then the payload will be routed to the default pipelines. When `silent` is used the error is not logged. If not supplied, `propagate` is used.

### Context Inference

The routing connector supports OTTL context inference, allowing you to write clearer and more maintainable routing conditions using context-qualified paths. This is the recommended approach for specifying routing conditions.

```yaml
# Recommended: Use context-qualified paths for clarity
- condition: resource.attributes["env"] == "prod"
  pipelines: [logs/prod]
- condition: span.attributes["http.method"] == "GET"
  pipelines: [traces/http]
- condition: log.severity_text == "ERROR"
  pipelines: [logs/errors]
```

This approach makes it immediately clear which attributes you're accessing without needing a separate `context` field.

> **Supported context-qualified paths:**

Reviewer: ```suggestion
### Supported contexts
```

| Context | Path prefix | Example |
|---------|-------------|---------|
| Resource | `resource.` | `resource.attributes["service.name"]` |
| Span | `span.` | `span.attributes["http.method"]` |
| Log | `log.` | `log.body`, `log.attributes["level"]` |
| Metric | `metric.` | `metric.name` |
| Datapoint | `datapoint.` | `datapoint.attributes["host"]` |

> | Otelcol | `otelcol.` | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |

Reviewer: ```suggestion
| Context     | Path prefix  | Example                                                                          |
|-------------|--------------|----------------------------------------------------------------------------------|
| [Resource]  | `resource.`  | `resource.attributes["service.name"]`                                            |
| [Span]      | `span.`      | `span.attributes["http.method"]`                                                 |
| [Log]       | `log.`       | `log.body`, `log.attributes["level"]`                                            |
| [Metric]    | `metric.`    | `metric.name`                                                                    |
| [Datapoint] | `datapoint.` | `datapoint.attributes["host"]`                                                   |
| [OtelCol]   | `otelcol.`   | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |

[resource]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlresource/README.md
[span]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlspan/README.md
[metric]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlmetric/README.md
[datapoint]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottldatapoint/README.md
[log]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottllog/README.md
[otelcol]: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlotelcol/README.md
```

  - `move`: Matched data is moved to the target pipeline(s) and removed from subsequent route evaluation. This is the default behavior.
  - `copy`: Matched data is copied to the target pipeline(s) but remains available for evaluation by subsequent routes. This allows the same data to be routed to multiple pipelines.
- `table.pipelines (required)`: the list of pipelines to use when the routing condition is met.
- `default_pipelines (optional)`: contains the list of pipelines to use when a record does not meet any of specified conditions.
- `error_mode (optional)`: determines how errors returned from OTTL statements are handled. Valid values are `propagate`, `ignore` and `silent`. If `ignore` or `silent` is used and a statement's condition has an error then the payload will be routed to the default pipelines. When `silent` is used the error is not logged. If not supplied, `propagate` is used.

### Context Inference

The routing connector supports OTTL context inference, allowing you to write clearer and more maintainable routing conditions using context-qualified paths. This is the recommended approach for specifying routing conditions.

```yaml

> # Recommended: Use context-qualified paths for clarity

Reviewer: ```suggestion
```

The following settings are available:

- `table (required)`: the routing table for this connector.

> - `table.condition`: the routing condition provided as the [OTTL] condition. Required if `table.statement` is not provided. Required for `request` context. Use context-qualified paths (e.g., `resource.attributes["key"]`, `span.attributes["key"]`) to automatically infer the context (see [Context Inference](#context-inference)).

Reviewer: Let's ask users to use qualified paths here, similar to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/48330/changes#diff-6a46f242e342c701ac32dc243603396800161de24c34ab4faff72c665216190fR40.

- `table.statement`: the routing condition provided as the [OTTL] statement. Required if `table.condition` is not provided. May not be used for `request` context. Generally `condition` is preferred since it is more terse.

> - `table.context (optional)`: the [OTTL Context] in which the condition/statement will be evaluated. Only `resource`, `span`, `metric`, `datapoint`, `log`, and `request` are supported. In most cases, you should omit this field and use context-qualified paths instead, which allows the context to be inferred automatically. If specified, takes precedence over inference. Defaults to `resource` when neither explicit context nor qualified paths are provided.

Reviewer: I think we should call out the `request` "context" deprecation here.

### Context Inference

The routing connector supports OTTL context inference, allowing you to write clearer and more maintainable routing conditions using context-qualified paths. This is the recommended approach for specifying routing conditions.

```yaml
# Recommended: Use context-qualified paths for clarity
- condition: resource.attributes["env"] == "prod"
  pipelines: [logs/prod]
- condition: span.attributes["http.method"] == "GET"
  pipelines: [traces/http]
- condition: log.severity_text == "ERROR"
  pipelines: [logs/errors]
```

This approach makes it immediately clear which attributes you're accessing without needing a separate `context` field.

**Supported context-qualified paths:**

| Context | Path prefix | Example |
|---------|-------------|---------|
| Resource | `resource.` | `resource.attributes["service.name"]` |
| Span | `span.` | `span.attributes["http.method"]` |
| Log | `log.` | `log.body`, `log.attributes["level"]` |
| Metric | `metric.` | `metric.name` |
| Datapoint | `datapoint.` | `datapoint.attributes["host"]` |

> | Any (requires explicit context) | `otelcol.` | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |

Reviewer: ```suggestion
| Otelcol | `otelcol.` | `otelcol.client.metadata["X-Tenant"][0]`, `otelcol.grpc.metadata["x-tenant"][0]` |
```

> The `otelcol.client.metadata` and `otelcol.grpc.metadata` paths provide access to incoming HTTP and gRPC request metadata respectively, and are valid in all signal contexts. **Important:** Context inference does not work for `otelcol.*` paths — you must set the `context` field explicitly (e.g., `context: resource`). Note that gRPC lowercases all metadata keys.

Reviewer: ```suggestion
The `otelcol.client.metadata` and `otelcol.grpc.metadata` paths provide access to incoming HTTP and gRPC request metadata respectively, and are valid in all signal contexts.
```
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44762/changes#r3233998184

> **Using explicit context:** If you prefer, you can still set the `context` field explicitly. This can be useful when working with unqualified paths:

Reviewer: ```suggestion
```

Let's not suggest this as the goal is to deprecate contextless paths.

## Examples

> [!NOTE]
> The examples below use context-qualified paths, which is the recommended configuration style. The explicit `context` field is still supported for backward compatibility but is no longer the primary documentation style. See [Context Inference](#context-inference) for details.

> ### Route traces using context-qualified paths

Reviewer: If all examples are using context-qualified paths, we probably don't need to add this extra one as it doesn't show any specific case.

```yaml
receivers:
  otlp:

exporters:
  file/prod:
    path: ./prod.json
  file/debug:
    path: ./debug.json
  file/other:
    path: ./other.json

connectors:
  routing:
    default_pipelines: [traces/other]
    table:
      - condition: resource.attributes["deployment.environment"] == "production"
        pipelines: [traces/prod]
      - condition: span.attributes["debug"] == true
        pipelines: [traces/debug]

service:
  pipelines:
    traces/in:
      receivers: [otlp]
      exporters: [routing]
    traces/prod:
      receivers: [routing]
      exporters: [file/prod]
    traces/debug:
      receivers: [routing]
      exporters: [file/debug]
    traces/other:
      receivers: [routing]
      exporters: [file/other]
```

### Route logs based on tenant (explicit context required)

> When conditions use `otelcol.client.metadata` or `otelcol.grpc.metadata` paths, the `context` field must be set explicitly because these paths cross-cut all signals and cannot be inferred. This is the primary case where an explicit `context` is needed.

Reviewer: ```suggestion
### Route logs based on tenant

```

We actually can use `otelcol` statements without explicitly set the `context` key.

> ## Examples

Reviewer: All examples should be updated/use qualified paths and inferred contexts, as it's the format we want to guide users to. We can add a note similar to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/transformprocessor/README.md?plain=1#L21 linking to the previous docs as well, and maybe a single example where setting the context manually makes sense, such as the examples listed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/filterprocessor/README.md#context-inference).

  routing:
    default_pipelines: [logs/other]
    table:
      # otelcol.client.metadata accesses HTTP request headers; use otelcol.grpc.metadata for gRPC.
      # Note: gRPC lowercases all metadata keys.
      - context: resource
        condition: otelcol.client.metadata["X-Tenant"][0] == "acme"
        pipelines: [logs/acme]
      - context: resource
        condition: otelcol.client.metadata["X-Tenant"][0] == "ecorp"

>         pipelines: [logs/ecorp]

Reviewer: ```suggestion
      - condition: otelcol.client.metadata["X-Tenant"][0] == "acme"
        pipelines: [logs/acme]
      - condition: otelcol.client.metadata["X-Tenant"][0] == "ecorp"
        pipelines: [logs/ecorp]
```

connectors:
  routing:
    table:
      - condition: log.severity_number < SEVERITY_NUMBER_ERROR
        pipelines: [logs/cheap]

>       - context: resource
>         condition: otelcol.client.metadata["X-Tenant"][0] == "acme"
>         pipelines: [logs/acme]
>       - context: resource
>         condition: otelcol.client.metadata["X-Tenant"][0] == "ecorp"
>         pipelines: [logs/ecorp]

Reviewer: ```suggestion
      - condition: otelcol.client.metadata["X-Tenant"][0] == "acme"
        pipelines: [logs/acme]
      - condition: otelcol.client.metadata["X-Tenant"][0] == "ecorp"
        pipelines: [logs/ecorp]
```

  match_once: true
  default_pipelines: [ logs/default ]
  table:

>     - condition: attributes["region"] == "west"

Reviewer: ```suggestion
```

    - condition: attributes["region"] == "east"
       pipelines: [ logs/east ]
    - condition: attributes["region"] == "west"

>     - condition: resource.attributes["env"] == "prod"

Reviewer: I think this new content should replace the above conditions without context's prefix. It seems it got appended to the list instead.

Author: It is now removed above
